skimage/segmentation/_join.py

```diff
@@ -114,20 +116,29 @@ def relabel_sequential(label_field, offset=1):
     >>> relab
     array([5, 5, 6, 6, 7, 9, 8])
     """
+    offset = int(offset)
+    if offset <= 0:
+        raise ValueError("Offset must be strictly positive.")
+    if int(label_field.min()) < 0:
+        raise ValueError("Cannot relabel array that contains negative values.")
     m = label_field.max()
-    if not np.issubdtype(label_field.dtype, np.signedinteger):
+    if not np.issubdtype(label_field.dtype, np.integer):
         new_type = np.min_scalar_type(int(m))
         label_field = label_field.astype(new_type)
         m = m.astype(new_type)  # Ensures m is an integer
     labels = np.unique(label_field)
     labels0 = labels[labels != 0]
-    if m == len(labels0):  # nothing to do, already 1...n labels
+    required_type = np.min_scalar_type(offset + len(labels0))
+    if np.dtype(required_type).itemsize > np.dtype(label_field.dtype).itemsize:
+        label_field = label_field.astype(required_type)
+    new_labels0 = np.arange(offset, offset + len(labels0))
+    if np.all(labels0 == new_labels0):
         return label_field, labels, labels
-    forward_map = np.zeros(m + 1, int)
-    forward_map[labels0] = np.arange(offset, offset + len(labels0))
+    forward_map = np.zeros(int(m + 1), dtype=label_field.dtype)
```

Comment:
Btw, I replaced `m + 1` with `int(m + 1)` to fix an issue when `m` is of type `np.uint64`.

Is this intended behavior or a numpy bug?
```python
>>> (np.uint64(5) + 1).dtype
dtype('float64')
```

Reply:
yikes! I think actually I came across this and the answer was that it is indeed intended, because (a) NumPy's type promotion rules do not depend on the input values, only on the types, and (b) they are supposed to be safe in the sense of being able to represent the result, and `float64` is the only thing that can represent anything with a Python int, since they are unbounded.

At least, that's my memory of it. Perhaps @stefanv has more details. But, either way, thanks for the fix!

Reply:
What is this sorcery!!!!

Reply:
Looks like the argument Juan outlines is being followed here.  Counter-intuitive, but probably correct?  Same as `(np.uint8(3)/2).dtype`.